****

# <center> <b> <span style="color:orange;"> Python Proficiency for Scientific Computing and Data Science (PyPro-SCiDaS)  </span> </b></center>

### <center> <b> <span style="color:green;">An Initiation to Programming using Python (Init2Py) </span> </b></center>
    


****

# <center> <b> <span style="color:blue;">Lecture 1: Variables and assignments </span> </b></center>


****



### <left> <b> <span style="color:brown;">Instructeur : </span> </b></left>[Yaé Ulrich Gaba](https://github.com/gabayae)




>    **Summary:** As soon as we have *data types*, we need *variables* to store the data. In reality, Python does not offer the concept of a variable, but rather that of an *object reference*. As long as the object is immutable (like integers, floats, etc.), there is no significant difference. In this notebook, we discuss the main points surrounding the use of variables in Python.

****

## 0. Introduction

In Python, a variable is a fundamental programming construct that facilitates the storage and management of information in the computer's memory. It is assigned a symbolic name, which serves as a reference to the data stored at a particular location in memory. This symbolic name enables you to access and modify the stored information throughout your code with ease.

Variables in Python are defined using the assignment operator =. You can assign a value to a variable directly or assign the value from another variable, making Python flexible in how it handles data. Rather than containing the data themselves, variables in Python act as references to objects in memory. This reference-based approach allows for efficient memory management and provides the flexibility to interact with and modify the data dynamically.

Python's dynamic typing system means that variables can be reassigned to different types of data during program execution. This dynamic nature, combined with the reference model, enhances the language's capability to handle various data types and operations seamlessly. Additionally, Python's garbage collection system manages memory allocation and deallocation, further optimizing performance and resource utilization.

## 1. Define variable

In Python, there are four fundamental ways to define a variable, each suited to different purposes and contexts:

1. **Direct Assignment**: This is the most straightforward method where a variable is assigned a specific value directly. This method is used for initializing variables with known values that will be used later in the code.


2. **Multiple Assignment**: This technique allows you to assign the same value to multiple variables simultaneously. It is useful for initializing several variables with the same value efficiently.

3. **Parallel Assignment**: This method involves defining multiple variables in a single statement with different values. It is effective for initializing several variables at once with distinct values.


4. **Assignment from Expressions**: Variables can be defined based on expressions involving other variables. This method allows you to create variables that represent calculations or transformations of existing variables.


Each of these methods provides flexibility in how you manage and utilize variables in your code, catering to different scenarios and requirements.


### 1.0. Through direct assignment


```python
y = 4.8134  # Defines a variable named y and assigns it the value 4.8134
salutation = "How are you ?"  # Defines a variable named salutation and assigns it the value "How are you?"
```

This code snippet illustrates the process of creating variables in Python by assigning them specific values. The variable `y` is assigned a floating-point number, while the variable `salutation` is assigned a string value. In Python, this assignment operation creates a reference between the variable name and the data, allowing you to reuse and manipulate the data through the variable name in subsequent parts of your code.

In [47]:
import math
y = math.pi  # Defines a variable named y and assigns it the value 4.8134
salutation = "How are you Walter?"  # Defines a variable named salutation and assigns it the value "How are you?"
country ="Zimbabwe"
x = 56

To display the values of the three defined variables, we use the `print()` function—a function that we will discuss in more detail when we cover function objects in Python. The `print()` function is essential in Python for outputting data to the console, allowing you to see the current state of your variables and debug your code effectively. This function can take multiple arguments, meaning you can print several variables at once, and it automatically converts them to their string representation if necessary.

In [49]:
print(f"{salutation} from {country}. Do you like {y}")

How are you Walter? from Zimbabwe. Do you like 3.141592653589793


To display the three values on the same line, you use a single `print()` function, separating the variables with commas. The commas in the print() function automatically insert spaces between the values, making it easy to format your output in a readable manner. This approach allows you to output multiple pieces of data in a single line without needing to manually concatenate strings or add spaces.

In [50]:
print(f"{salutation} from {country}. Do you like {y}")

How are you Walter? from Zimbabwe. Do you like 3.141592653589793


### 1.1. Through multiple assignment

The examples presented fall under what we call `direct assignment`. A `multiple assignment` is a specific case of `direct assignment` where the same value is assigned to multiple variables in a single line of code. This technique is useful when you need to initialize several variables with the same starting value efficiently, keeping your code concise and readable.


```python
x = y = 7  # x and y are both assigned the value 7 simultaneously.
```

In this line of code, the variables `x` and `y` are assigned the same value of $7$ at the same time. This is an example of a `multiple assignment`, where a single value is efficiently assigned to multiple variables in one statement. This technique is particularly useful when initializing variables with the same initial value, ensuring consistency across your code.

In [51]:
x = y = "How is Aims Rwanda"  # x and y are both assigned the value 7 simultaneously.
print(x)
print("=======================")
print(y)

How is Aims Rwanda
How is Aims Rwanda


### 1.2. Parallel assignment



A `parallel assignment` involves defining multiple variables using a single equals sign. This technique allows you to assign different values to several variables simultaneously in a single line of code. It enhances code readability and efficiency, especially when you need to initialize multiple variables at once. Example:


```python
x, y = 4, 8.33  # Defines two variables, x and y, with values 4 and 8.33 respectively.
```

In this line of code, `x` is assigned the value 4, and `y` is assigned the value 8.33 simultaneously. This is an example of `parallel assignment`, where multiple variables are defined in a single statement. This approach is particularly useful for initializing or updating several variables in a concise and organized manner.

In [52]:
x, y = 10, 89.33  # Defines two variables, x and y, with values 10 and 89.33 respectively.

In [54]:
print(f"Value of x is {x} and the value of y is {y}")

Value of x is 10 and the value of y is 89.33


In [57]:
x, y, z = 10, 89.33, 12  # Defines two variables, x,y and z with values 10 and 89.33 respectively, 

print(f"x= {x}\ny={y}\nz={z}")

x= 10
y=89.33
z=12


### 1.3. Based on other variables

In [58]:
#Define a variable based on other variables
z1 = x + y  # Defines the variable named z1 and assigns it the sum of variables x and y
z2 = x + 5  # Defines the variable named z2 by adding 5 to the value of x
z3 = 2 * y  # Defines the variable named z3 by multiplying the value of y by 2

print(z1,z2,z3)

99.33 15 178.66


<left> <b> <span style="color:red;">Assignment is not comparison!</span> </b></left>
It is important to note that the assignment operator `=` does not have the same meaning as the equality symbol `=` in mathematics. For example, the assignment operator is not symmetric, while the equality symbol is: attempting to swap the order of elements in an assignment statement will inevitably result in an error in the interpreter:


In [13]:
# Error
128 = a

SyntaxError: cannot assign to literal (<ipython-input-13-19d32c7debe7>, line 2)

In [ ]:
a = 128

This brings us to briefly discuss permissible variable names in Python.

#### <left> <b> <span style="color:brown;">Naming Conventions</span> </b></left>


Naming conventions for different elements of code are important because they provide additional information to developers about the nature of certain attributes or variables. The conventions for variable names are as follows:

   - Reserved keywords such as `if`, `else`, etc., cannot be used as variable names.
   - Variable names can start with `_`, `$`, or a letter.
   - Variable names can be in lowercase or uppercase.
   - Variable names cannot start with a digit.
   - White spaces are not allowed in variable names.

<left> <b> <span style="color:brown;"> A good programmer naturally strives to choose the most meaningful variable names possible. </span> </b></left>


In Python, there are **33** reserved keywords, and the list is provided below:


|             |                  |                  |                  |                  |
|-------------|------------------|------------------|------------------|------------------|
|``and``      | ``elif``         | ``if``           | ``or``           | ``yield``        |
|``as``       | ``else``         | ``import``       | ``pass``         |                  |
|``assert``   | ``except``       | ``in``           | ``raise``        |                  |  
|``break``    | ``False``        | ``is``           | ``return``       |                  |
|``class``    | ``finally``      | ``lambda``       | ``True``         |                  |
|``continue`` | ``for``          | ``None``         | ``try``          |                  |
|``def``      | ``from``         | ``nonlocal``     | ``while``        |                  |
|``del``      | ``global``       | ``not``          | ``with``         |                  |

To get the list of reserved keywords in Python, you can use the `keyword` module. Here’s how you can do it:

1. **Import the `keyword` module**:
   ```python
   import keyword
   ```

2. **Use the `keyword.kwlist` attribute to get the list of keywords**:
   ```python
   print(keyword.kwlist)
   ```

3. **To check if a specific word is a keyword**:
   ```python
   print(keyword.iskeyword('if'))  # Returns True
   print(keyword.iskeyword('my_var'))  # Returns False
   ```

The `keyword.kwlist` attribute returns a list of all reserved keywords in Python, and the `keyword.iskeyword()` function checks if a given string is a keyword.

In [ ]:
import keyword
print(keyword.kwlist)

In [ ]:
print(keyword.iskeyword('if'))  # Returns True
print(keyword.iskeyword('my_var'))  # Returns False

**Note**: Python is case-sensitive, so variable names `Age` and `age` are considered distinct. Depending on the language, there is a [naming convention](https://en.wikipedia.org/wiki/Naming_convention_(programming)#Python_and_Ruby) that is recommended:

- `UpperCamelCase` for class names;

- `CAPITALIZED_WITH_UNDERSCORES` for constants;

- `lowercase_separated_by_underscores` or `snake_case` for other variables.


****
**A Fundamental Exercise: Swapping the Contents of Two Variables**

> Let's assume that the variables `x` and `y` have the values of integers $\alpha$ and $\beta$ respectively. The goal is to swap the contents of these two variables.

>   - a. First Method: Propose a method that uses an auxiliary variable `tmp`.
       ```python
          tmp = x
          x = y
          y = tmp
       ```
>   - b. Second Method: Execute the following sequence of instructions:
       ```python
          x = x + y; y = x - y; x = x - y  
       ```
       
>   - c. Third Method (the most "Pythonic"): Use parallel assignment.
      ```python
         x, y = y, x
      ```   
****

Example:


In [59]:
# First Method

x = 5.7
y = 10
print(x, y)
print("================================================")
print('\t')

tmp = x
x = y
y = tmp
print(x, y)

# Second Method

x = 7
y = 9
print(x, y)
print("================================================")
print('\t')

x = x + y; y = x - y; x = x - y
print(x, y)

# Third Method

x = 90
y = 15
print(x, y)
print("================================================")
print('\t')

x, y = y, x
print(x, y)


5.7 10
	
10 5.7
7 9
	
9 7
90 15
	
15 90


It's worth noting that to delete a variable in Python, you can use the `del` function. This function removes the variable from the current namespace, effectively deleting it and freeing up any resources it was using. For example:

```python
x = 10  # Define a variable x
del x   # Delete the variable x
```

After executing `del x`, the variable `x` will no longer exist in the current scope, and attempting to access it will result in a `NameError`.

In [60]:
x = 19.999  # Define a variable x
print(x)

del x   # Delete the variable x
print(x)

19.999


NameError: name 'x' is not defined

## 2. Type of a Variable

In Python, the type of a variable refers to the kind of data it holds, such as integers, floating-point numbers, strings, or more complex data structures. Python dynamically assigns the type based on the value assigned to the variable. This type can be determined at any point in the code using the `type()` function.

The type of a variable corresponds to its nature. There are many types of variables (integer, real number, strings, etc.). The most commonly encountered types of variables are integers (`int`), real numbers (`float`), and strings (`str`).

**The basic types include:**

- **None** (nothing)
- **String types:** `str`
   - Enclosed in (single, double, or triple) quotes `'` or `"`: `'Calvin'`, `"Calvin'n'Hobbes"`, `'''Two\nlines'''`, `"""'Why?' he asked."""`
   - Conversion: `str(3.2)`
- **Numeric types:**
   - **Booleans** `bool` (true/false): `True`, `False`
   - **Integers** `int` (no explicit limit value, corresponds to at least C's long type): `-2`, `int(2.1)`, `int("4")`
   - **Reals** `float`
   - **Complex** `complex`: `1+2j`, `5.1j`, `complex(-3.14)`, `complex('j')`
- **Iterable objects:**
   - **Lists** `list`: `['a', 3, [1, 2], 'a']`
   - **Immutable lists** `tuple`: `(2, 3.1, 'a', [])` (depending on the usage, parentheses are not always required)
   - **Keyed lists** `dict`: `{'a':1, 'b':[1, 2], 3:'c'}`
   - **Unordered sets of unique elements** `set`: `{1, 2, 3, 2}`

### 2.0. None (nothing)
The `None` type represents the absence of a value or a null value in Python. It is often used to signify that a variable has no value assigned to it or to indicate the end of a list, function, or loop.

**Example:**
```python
x = None
```

In [61]:
x = True
print(x, type(x))

True <class 'bool'>


### 2.1. String Types (`str`)
Strings in Python are sequences of characters enclosed in quotes. They can be defined using single (`'`), double (`"`), or triple quotes (`'''` or `"""`). Triple quotes allow for multi-line strings.

**Examples:**
```python
name = 'Calvin'
quote = "Calvin'n'Hobbes"
multi_line = '''Two
lines'''
```

In [62]:
name = 'Walter'
quote = "Walter'n'Rugby"
multi_line = 'Two'

In [63]:
print(name)

Walter


In [64]:
print(quote)

Walter'n'Rugby


In [65]:
type(name)

str

In [66]:
print(name, type(name), '\n')
print('====================================')

print(quote, type(quote), '\n')
print('====================================')

print(multi_line, type(multi_line), '\n')
print('====================================')

Walter <class 'str'> 

Walter'n'Rugby <class 'str'> 

Two <class 'str'> 



### 2.2. Numeric Types

#### 2.2.0. Booleans (`bool`)
Booleans represent one of two values: `True` or `False`. They are often used in conditional statements to determine the flow of a program.

**Examples:**
```python
is_active = True
has_permission = False
```

In [67]:
is_active = True
has_permission = False


print(is_active, type(is_active), '\n')
print('====================================')

print(has_permission, type(has_permission), '\n')


True <class 'bool'> 

False <class 'bool'> 



####  2.2.1. Integers (`int`)
Integers are whole numbers without a fractional component. In Python, integers can be of arbitrary precision, meaning they can be as large as the memory allows.

**Examples:**
```python
age = 25
negative_number = -42
```

In [68]:
age = 1000
negative_number = -42

print(age, type(age), '\n')
print('====================================')
print(negative_number, type(negative_number))

1000 <class 'int'> 

-42 <class 'int'>


####  2.2.2. Reals (`float`)
Floating-point numbers (floats) are numbers with a decimal point. They are used to represent real numbers in Python.

**Examples:**
```python
pi = 3.14159
temperature = -2.5
```

In [69]:
pi = 10.6
temperature = -2.5


print(pi, type(pi), '\n')
print('====================================')
print(temperature, type(temperature))

10.6 <class 'float'> 

-2.5 <class 'float'>


#### 2.2.3. Complex (`complex`)
Complex numbers in Python consist of a real part and an imaginary part. They are represented by `a + bj`, where `a` is the real part and `b` is the imaginary part.

**Examples:**
```python
z = 1 + 2j
w = complex(3, -4)
```

In [70]:
z = 10 + 22j
w = complex(3, -4)

print(z, type(z))
print('====================================')
print(w, type(w))

(10+22j) <class 'complex'>
(3-4j) <class 'complex'>


### 2.3. Iterable Objects

#### 2.3.0. Lists (`list`)
A list is an ordered collection of items that can be of different types. Lists are mutable, meaning their contents can be changed after creation.

**Examples:**
```python
fruits = ['apple', 'banana', 'cherry']
mixed = [1, 'two', 3.0, [4, 5]]
```

In [71]:
fruits = ['apple', 'banana', 'cherry','watermelon']
mixed = [1, 'two', 3.0, [4, 5]]


print(fruits, type(fruits))
print('====================================')
print(mixed, type(mixed))

['apple', 'banana', 'cherry', 'watermelon'] <class 'list'>
[1, 'two', 3.0, [4, 5]] <class 'list'>


#### 2.3.1. Immutable Lists (`tuple`)
A tuple is similar to a list but is immutable, meaning its contents cannot be changed after creation. Tuples are often used to store collections of related data.

**Examples:**
```python
coordinates = (10.5, 20.8)
colors = ('red', 'green', 'blue')
```

In [72]:
coordinates = (10.5, 50,20.8)
colors = ('red', 'green', 'blue', 'yellow')

print(coordinates, type(coordinates))
print('====================================')
print(colors, type(colors))

(10.5, 50, 20.8) <class 'tuple'>
('red', 'green', 'blue', 'yellow') <class 'tuple'>


#### 2.3.2. Keyed Lists (`dict`)
A dictionary is a collection of key-value pairs, where each key is associated with a value. Dictionaries are mutable and allow for fast lookup of values based on their keys.

**Examples:**
```python
person = {'name': 'Alice', 'age': 30}
inventory = {'apples': 10, 'bananas': 20}
```

In [73]:
person = {'name': 'Alice', 'age': 30, 'car': 'Mustang'}
inventory = {'apples': 10, 'bananas': 20, 'orange': 60}


print(person, type(person))
print('====================================')
print(inventory, type(inventory))

{'name': 'Alice', 'age': 30, 'car': 'Mustang'} <class 'dict'>
{'apples': 10, 'bananas': 20, 'orange': 60} <class 'dict'>


#### 2.3.3. Unordered Sets of Unique Elements (`set`)
A set is an unordered collection of unique elements. Sets are useful for membership tests and eliminating duplicate entries.

**Examples:**
```python
unique_numbers = {1, 2, 3, 2}
letters = {'a', 'b', 'c', 'a'}
```

In [74]:
unique_numbers = {1, 2, 3, 2}
letters = {'a', 'b', 'c', 'a'}

print(unique_numbers, type(unique_numbers))
print('====================================')
print(letters , type(letters ))


{1, 2, 3} <class 'set'>
{'c', 'b', 'a'} <class 'set'>


### 2.4. Dynamic Typing in Python

Python is a dynamically typed language, meaning that the type of a variable is determined at runtime rather than at compile time. In Python, you don't need to declare the type of a variable when you create it. Instead, the type is inferred based on the value assigned to the variable. This allows for more flexibility but also requires careful handling to avoid type-related errors.

#### Key Characteristics of Dynamic Typing:

- **No Type Declaration**: You simply assign a value to a variable, and Python automatically knows what type it is.
    ```python
    x = 10        # x is an integer
    x = "hello"   # Now, x is a string
    ```
- **Type Flexibility**: The type of a variable can change over its lifetime. You can reassign a variable to a value of a different type without any issues.
    ```python
    y = 3.14      # y is initially a float
    y = True      # Now, y is a boolean
    ```
- **Memory Management**: Python handles memory management automatically. When you reassign a variable to a new value, the previous value is discarded if it’s no longer referenced elsewhere in the code.

#### Pros and Cons of Dynamic Typing:

- **Pros**:
  - **Flexibility**: You can write more general-purpose code since the type is not fixed.
  - **Ease of Use**: Less boilerplate code, as there is no need for explicit type declarations.
  
- **Cons**:
  - **Type-Related Errors**: Since types are determined at runtime, it’s possible to encounter errors if the wrong type is used in an operation.
  - **Performance**: Dynamic typing can be slower than static typing because type checks are done at runtime.

#### Example:
```python
# Initially, 'data' is an integer
data = 100

# Now, 'data' is a string
data = "Dynamic Typing"

# And now 'data' is a list
data = [1, 2, 3]

# Python handles these changes without any issues
```

### 2.5.  Coercion in Python

**Coercion** in Python refers to the automatic conversion of one data type to another during operations that involve different types. Python is designed to handle these type conversions in a way that makes the language easier to use and reduces the need for manual type casting.

#### Key Points About Coercion:

- **Implicit Coercion**: Python automatically converts one data type to another when necessary to perform an operation. This usually happens in arithmetic operations involving different types, like an integer and a float.
  - For example, if you add an integer to a float, Python will convert the integer to a float before performing the addition.
  
- **Explicit Coercion**: While Python handles many conversions automatically, you can also manually convert types using built-in functions like `int()`, `float()`, `str()`, etc. This is known as explicit type casting.

#### 2.5. 0. Implicit Coercion Example:

```python
# Adding an integer and a float
x = 5        # int
y = 3.2      # float

# Python automatically converts 'x' to a float before performing the addition
result = x + y

print(result)  # Output: 8.2 (float)
```

In the example above, Python automatically converts the integer `5` to a float `5.0` to perform the addition with the float `3.2`, resulting in a float `8.2`.

In [75]:
# Adding an integer and a float
x = 5        # int
y = 3.2      # float

# Python automatically converts 'x' to a float before performing the addition
result = x + y




print(x, type(x), '\n')
print('====================================')
print(y, type(y),'\n')
print('====================================')
print(result, type(result))  # Output: 8.2 (float)

5 <class 'int'> 

3.2 <class 'float'> 

8.2 <class 'float'>


#### 2.5.1. Explicit Coercion Example:

```python
# Converting a float to an integer
a = 7.9
b = int(a)  # Explicit coercion using the int() function

print(b)  # Output: 7 (integer, with the decimal part truncated)
```

Here, the float `7.9` is explicitly converted to the integer `7` using the `int()` function, which removes the fractional part.

In [36]:
a = 7.9
b = int(a)  # Explicit coercion using the int() function

print(b)  # Output: 7 (integer, with the decimal part truncated)

7


In [76]:
a = int(math.pi)
a

3


#### 2.5.2. Common Coercion Scenarios:

- **String to Integer/Float**: When you need to convert a string containing numeric characters to an integer or float.
  ```python
  num_str = "123"
  num_int = int(num_str)   # Converts to integer 123
  num_float = float(num_str)  # Converts to float 123.0
  ```

- **Integer/Float to String**: When you need to concatenate a number with a string.
  ```python
  age = 25
  message = "I am " + str(age) + " years old."
  ```

- **Boolean to Integer**: `True` is coerced to `1` and `False` to `0` in numeric operations.
  ```python
  result = True + 2   # Output: 3 (1 + 2)
  ```

#### 2.5.3. Pros and Cons of Coercion:

- **Pros**:
  - Simplifies code by reducing the need for explicit type conversions.
  - Makes the language more intuitive and user-friendly.

- **Cons**:
  - Can lead to unexpected results if the automatic type conversion doesn't align with the programmer's intent.
  - Potentially hides bugs related to incorrect data types.

Coercion in Python allows for smoother and more intuitive operations involving different data types. While it adds convenience, it's important to understand how and when Python performs these conversions to avoid unexpected behaviors.

## 3. Methods associated with variables

In Python, every variable is linked to a variety of attributes and methods that define its behavior and interactions. These methods are functions that are built into the variable's type and allow you to perform various operations on the variable. For example, methods can help you manipulate strings, perform mathematical operations, or interact with lists and dictionaries.

The `dir()` function is useful for exploring these methods and understanding what operations are available for a given variable. By calling `dir()` on a variable, you get a list of all its attributes and methods, including those inherited from its type. This can be particularly helpful for discovering how to use a variable's methods or for debugging.

Here's how you might use `dir()`:

```python
# Example with a string variable
text = "Hello, world!"
print(dir(text))

# Example with a list variable
numbers = [1, 2, 3, 4, 5]
print(dir(numbers))
```

In the examples above, `dir(text)` will list methods related to string operations such as `upper()`, `lower()`, and `split()`, while `dir(numbers)` will show methods related to list operations like `append()`, `remove()`, and `sort()`. This feature of Python makes it easier to explore and utilize the functionalities associated with different data types.

In [37]:
x = 2.5 # Définit une variable numérique x
y = 'my text' # Définit une variable en chaîne de caractères y.

Pour afficher l’ensemble des méthodes associées à chacune de ces variables, on fait :


In [77]:
print(dir(person))

['__class__', '__class_getitem__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__ior__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__or__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__ror__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


In [78]:
print(dir(fruits))

['__add__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


You can also type `variable_name.` followed by `TAB` in many interactive Python environments or integrated development environments (IDEs). This action triggers autocompletion, which helps you see and select available methods and attributes associated with that variable. For example:

1. Type `variable_name.` and press `TAB`.
2. A list of methods and attributes that can be used with `variable_name` will appear.

This feature is especially useful for exploring what operations you can perform on a variable and for quickly finding the right method without needing to remember exact method names.

In [ ]:
# Type x. followed by `TAB`

fruits.

To get help on a specific method in Python, you can use the `help()` function. The syntax is `help(variable_name.method_name)`, where `method_name` is the name of the method you are interested in. For example, if you have a numeric variable of type `float`, which includes a method named `conjugate`, you can obtain information about this method by running:

```python
print(help(x.conjugate))
```

This will display documentation about the `conjugate` method, including its purpose and usage, in the console or terminal.

In [79]:
print(help(x.conjugate))

Help on built-in function conjugate:

conjugate(...) method of builtins.int instance
    Returns self, the complex conjugate of any int.

None


To display help on all the functions associated with a variable `x`, you simply use:

```python
help(x)
```

This command will show the documentation for the type of the variable `x`, including all the methods and attributes available for that type.

In [80]:
print(help(x))

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil_

### 4. Arithmetic and Logical (Boolean) Operators

In Python, two major categories of operators are used to define variables and instructions: arithmetic operators and logical (boolean) operators.

- **Arithmetic Operators**: These operators perform common mathematical operations. They include addition, subtraction, multiplication, division, and others that are essential for numerical computations.

- **Logical (Boolean) Operators**: These operators are used for comparing values and evaluating logical expressions. A boolean value represents one of two possibilities: true or false. Boolean values result from evaluating logical expressions and are used to make decisions within a program, such as executing certain actions when specific conditions are met.

**Boolean values** are crucial for control flow in programming, allowing for conditional execution based on whether a condition evaluates to true or false.




#### 4.0. Arithmetic Operators

| **Operation**            | **Symbol**  | **Example**       |
|:-------------------------:|:-----------:|:-----------------:|
| Addition                  | `+`         | `x = 2 + 3`       |
| Subtraction               | `-`         | `z = x - y`       |
| Multiplication            | `*`         | `y = 3 * x`       |
| Real Division             | `/`         | `5 / 2 = 2.5`    |
| Integer Division          | `//`        | `5 // 2 = 2`     |
| Exponentiation            | `**`        | `x ** 2 = x * x` |
| Modulo (Remainder)        | `%`         | `17 % 3 = 2`     |
| Increment Addition        | `+=`        | `x += 4` (i.e., `x = x + 4`) |
| Increment Subtraction     | `-=`        | `x -= 4` (i.e., `x = x - 4`) |

---





#### 4.1. Logical Operators<

| **Operation**                | **Symbol** | **Description**                                | **Example**               |
|:----------------------------:|:----------:|:---------------------------------------------:|:--------------------------:|
| Logical AND                  | `and`      | Returns `True` if both operands are true     | `True and False` yields `False` |
| Logical OR                   | `or`       | Returns `True` if at least one operand is true| `True or False` yields `True`  |
| Logical NOT                  | `not`      | Returns `True` if the operand is false       | `not True` yields `False`      |
| Logical XOR (Exclusive OR)   | `^`        | Returns `True` if operands are different     | `True ^ False` yields `True`   |
| Logical equality             | `==`       | Returns `True` if both operands are equal    | `x == y`                    |
| Logical inequality           | `!=`       | Returns `True` if operands are not equal    | `x != y`                    |
| Less than                    | `<`        | Returns `True` if left operand is less than right operand | `x < y`                  |
| Greater than                 | `>`        | Returns `True` if left operand is greater than right operand | `x > y`                  |
| Less than or equal to        | `<=`       | Returns `True` if left operand is less than or equal to right operand | `x <= y`                 |
| Greater than or equal to     | `>=`       | Returns `True` if left operand is greater than or equal to right operand | `x >= y`                 |

---



To access the complete list of standard Python operators and their equivalent functions, see [this page](https://docs.python.org/3/library/operator.html). You can also refer to [this page](https://www.tutorialspoint.com/python/python_operators.htm) for some examples of standard operator usage.



### 5. User input ( the `input()` function)

In Python, the `input()` function is used to capture user input from the console. It pauses the program's execution and waits for the user to type something, which is then returned as a string. This input can be stored in a variable, allowing you to use the entered data later in your code.

#### 5.0. Basic Usage

The basic syntax for the `input()` function is:

```python
variable_name = input(prompt)
```

- **`prompt`**: This is an optional argument. It is a string that is displayed to the user, providing instructions or asking for specific input.
- **`variable_name`**: This is the variable that will store the value entered by the user.

#### Example

Here’s a simple example of using the `input()` function:

```python
name = input("Enter your name: ")
print("Hello, " + name + "!")
```

In this example:
- The program prompts the user to enter their name.
- The entered name is stored in the variable `name`.
- The program then greets the user using the name provided.

In [81]:
name = input("Enter your name: ")
print("Hello, " + name + "!")

Enter your name: Walter
Hello, Walter!


#### 5.1. Important Notes

- **Type Conversion**: Since `input()` always returns the input as a string, you may need to convert it to the appropriate type (e.g., `int`, `float`) depending on the context.

  ```python
  age = int(input("Enter your age: "))
  ```

- **Handling Errors**: When converting input, it’s important to handle potential errors, such as the user entering a non-numeric value when an integer is expected.

- **Security Considerations**: Be cautious when using `input()` in sensitive applications, as it can introduce security risks if the input is not properly validated or sanitized.

Using the `input()` function is a common way to make your Python programs interactive, enabling users to provide data that can be processed and utilized by the program.

#### 5.2.  Accepting User Inputs (as both integer and string)

**`input(prompt)`** prompts for and returns input as a string. Hence, if the user inputs a integer, the code should convert the string to an integer and then proceed.

In [82]:
a = input("Hello, \nHow are you?  ") # \n means new line


print("================================== \n")
print(type(a))

Hello, 
How are you?  Walter

<class 'str'>


In [41]:
try_something = input("Type something here and it will be stored in variable try_something \t")

print("================================== \n")
print(type(try_something))

Type something here and it will be stored in variable try_something 	10

<class 'str'>


In [42]:
number = input("Enter number: ")
name = input("Enter name: ")

print("\n")
print("Printing type of a input value")
print("================================== \n")
print("Type of number", type(number))
print("================================== \n")
print("Type of name", type(name))

Enter number: 4
Enter name: 4


Printing type of a input value

Type of number <class 'str'>

Type of name <class 'str'>


#### 5.3. `eval()` (accepting user inputs; only as integer)

The `eval()` function in Python can be used in conjunction with `input()` to evaluate a string as a Python expression. This can be particularly useful when you want to allow the user to input a mathematical expression or Python code directly and have it evaluated at runtime.

#### Basic Usage of `eval()`

The basic syntax for using `eval()` with `input()` is:

```python
result = eval(input(prompt))
```

- **`prompt`**: This is the text displayed to the user to guide them on what to input.
- **`result`**: This variable stores the output after evaluating the user input as a Python expression.

#### Example

Here’s an example where the user is allowed to input a mathematical expression, and `eval()` evaluates it:

```python
expression = input("Enter a mathematical expression: ")
result = eval(expression)
print("The result is:", result)
```

In [43]:
expression = input("Enter a mathematical expression: ")
result = eval(expression)
print("The result is:", result)

Enter a mathematical expression: 5
The result is: 5



In this example:
- The program prompts the user to enter a mathematical expression.
- The input is passed to `eval()`, which evaluates the expression.
- The result is then printed out.

#### Important Notes

- **Use with Caution**: The `eval()` function can be dangerous if used with untrusted input, as it will execute any code passed to it. This could potentially lead to security vulnerabilities, such as code injection attacks. It should only be used in safe, controlled environments where the input is trusted.
  
- **Valid Python Expressions**: The string passed to `eval()` must be a valid Python expression. If the string contains syntax errors or invalid operations, Python will raise an exception.

- **Alternative**: In many cases, using `int()` or `float()` for type conversion, or safely parsing and evaluating input without `eval()`, may be preferable for security reasons.

Using `eval()` with `input()` can be powerful for dynamic code evaluation, but it should be used responsibly to avoid unintended consequences.


### Practice Problem

In [44]:
# Accept one integer and one float number from the user and calculate the addition of both the numbers.
num1 = int(input("Enter first number: "))
num2 = float(input("Enter second number: "))

result=(num1+num2)
print("Final result is: ", result)

Enter first number: 5
Enter second number: 5
Final result is:  10.0


In [45]:
# Write code to get three numbers and add first 2 number and multiply with third number

num1 = int(input("Enter first number: "))   # converting input value to integer
num2 = int(input("Enter second number: "))  # converting input value to integer
num3 = int(input("Enter third number: "))   # converting input value to integer

print("\n")
print("First Number: ", num1)
print("Second Number: ", num2)
print("Third Number: ", num3)

result=(num1+num2)*num3
print("Final result is: ", result)

Enter first number: 5
Enter second number: 5
Enter third number: 5


First Number:  5
Second Number:  5
Third Number:  5
Final result is:  50


In [83]:
# Write a code to get four numbers:
#Step 1. Multiply first and fourth number
#Step 2. Divide second and third number
#Step3. Add Step 1 and Step 2 outputs.

num1 = int(input("Enter first number: "))
num2 = int(input("Enter second number: "))
num3 = int(input("Enter third number: "))
num4 = int(input("Enter fourth number: "))

print("\n")
print("First Number: ", num1)
print("Second Number: ", num2)
print("Third Number: ", num3)
print("Fourth number: ", num4)

result=(num1*num4)+(num2/num3)
print("Final result is: ", result)

Enter first number: 5
Enter second number: 5
Enter third number: 5
Enter fourth number: 5


First Number:  5
Second Number:  5
Third Number:  5
Fourth number:  5
Final result is:  26.0


In [84]:
# Let’s see how to accept float value from a user in Python.
# You need to convert user input to the float number using
# the **`float()`** function as we did for the integer value.
float_number = float(input("Enter float number: "))  # converting input value to float
print("\n")
print("input float number is: ", float_number)
print("type is:", type(float_number))

Enter float number: 5.7


input float number is:  5.7
type is: <class 'float'>
